usr/lib/common/template.c

```diff
@@ -638,25 +636,25 @@ CK_RV template_unflatten_withSize(TEMPLATE **new_tmpl, CK_BYTE *buf,
                 TRACE_ERROR("%s\n", ock_err(ERR_HOST_MEMORY));
                 return CKR_HOST_MEMORY;
             }
-            a2->type = a1->type;
+            a2->type = a1.type;
 
-            if ((a1->type == CKA_CLASS || a1->type == CKA_KEY_TYPE
-                 || a1->type == CKA_MODULUS_BITS
-                 || a1->type == CKA_VALUE_BITS
-                 || a1->type == CKA_CERTIFICATE_TYPE
-                 || a1->type == CKA_VALUE_LEN)
-                && a1->ulValueLen != 0) {
+            if ((a1.type == CKA_CLASS || a1.type == CKA_KEY_TYPE
+                 || a1.type == CKA_MODULUS_BITS
+                 || a1.type == CKA_VALUE_BITS
+                 || a1.type == CKA_CERTIFICATE_TYPE
+                 || a1.type == CKA_VALUE_LEN)
+                && a1.ulValueLen != 0) {
 
                 a2->ulValueLen = sizeof(CK_ULONG);
 
                 {
-                    CK_ULONG_32 *p32;
+                    CK_ULONG_32 p32;
                     CK_BYTE *pb2;
 
-                    pb2 = (CK_BYTE *) a1;
+                    pb2 = (CK_BYTE *) ptr;
                     pb2 += sizeof(CK_ATTRIBUTE_32);
-                    p32 = (CK_ULONG_32 *) pb2;
-                    attr_ulong_32 = *p32;
+                    memcpy(&p32, pb2, sizeof(p32));
+                    attr_ulong_32 = p32;
                 }
```

Comment:
can't we do all the {} above much better ?
attr_ulong_32 = *((CK_ULONG_32 *)(((CK_BYPTE *)(ptr)) + sizeof(CK_ATTRIBUTE_32)));
lot of parenthesis ...

Reply:
No, you cant cast an u8 pointer (alignment 1) to an u32 (alignment 4) and dereference it without address sanitizer complaining.

Reply:
What about endianness here?